src/mono/mono/mini/interp/transform.c

```diff
@@ -734,6 +734,18 @@ handle_branch (TransformData *td, int long_op, int offset)
 		g_assert_not_reached ();
 	/* Add exception checkpoint or safepoint for backward branches */
 	if (offset < 0) {
+
+		InterpMethod *rtm = td->rtm;
+		guint16 samplepoint_profiling = 0;
+		if (mono_jit_trace_calls != NULL && mono_trace_eval (rtm->method))
+			samplepoint_profiling |= TRACING_FLAG;
+		if (rtm->prof_flags & MONO_PROFILER_CALL_INSTRUMENTATION_SAMPLEPOINT_CONTEXT)
+			samplepoint_profiling |= PROFILING_FLAG;
+		if (samplepoint_profiling) {
+			interp_add_ins (td, MINT_PROF_SAMPLEPOINT);
```

Comment:
Could MINT_PROF_SAMPLEPOINT also function as a MINT_SAFEPOINT to reduce the size of the generated interpreter opcodes when both sampling and safepoints are active?

Not sure how important it is for your scenarios though

Reply:
ST wasm doesn't emit MINT_SAFEPOINT and so I wanted to keep them separated. This profiler doesn't really work for MT anyway.